src/StorageRent.sol

```diff
@@ -0,0 +1,626 @@
+     *
+     * @param fid   The fid that will receive the credit.
+     * @param units Number of storage units to credit.
+     */
+    function credit(uint256 fid, uint256 units) external onlyOperator whenNotDeprecated {
+        if (units == 0) revert InvalidAmount();
+        if (rentedUnits + units > maxUnits) revert ExceedsCapacity();
+
+        rentedUnits += units;
+        emit Rent(msg.sender, fid, units);
+    }
+
+    /**
+     * @notice Credit multiple fids with free storage units. Only callable by operator.
+     *
+     * @param fids  An array of fids.
+     * @param units Number of storage units per fid.
+     */
+    function batchCredit(uint256[] calldata fids, uint256 units) external onlyOperator whenNotDeprecated {
+        if (units == 0) revert InvalidAmount();
+        uint256 totalUnits = fids.length * units;
+        if (rentedUnits + totalUnits > maxUnits) revert ExceedsCapacity();
+        rentedUnits += totalUnits;
+        for (uint256 i; i < fids.length; ++i) {
+            emit Rent(msg.sender, fids[i], units);
+        }
+    }
+
+    /**
+     * @notice Credit a continuous sequence of fids with free storage units. Only callable by operator.
+     *
+     * @param start Lowest fid in sequence (inclusive).
+     * @param end   Highest fid in sequence (inclusive).
+     * @param units Number of storage units per fid.
+     */
+    function continuousCredit(uint256 start, uint256 end, uint256 units) external onlyOperator whenNotDeprecated {
+        if (units == 0) revert InvalidAmount();
+        uint256 len = end - start;
+        uint256 totalUnits = len * units;
+        if (rentedUnits + totalUnits > maxUnits) revert ExceedsCapacity();
+        rentedUnits += totalUnits;
+        for (uint256 i; i < len; ++i) {
+            emit Rent(msg.sender, start + i, units);
+        }
+    }
+
+    /**
+     * @notice Force refresh the cached Chainlink ETH/USD price. Callable by admin and treasurer.
+     */
+    function refreshPrice() external {
+        if (!hasRole(ADMIN_ROLE, msg.sender) && !hasRole(TREASURER_ROLE, msg.sender)) revert Unauthorized();
+        _refreshPrice();
+    }
+
+    /**
+     * @notice Change the USD price per storage unit. Callable by admin and treasurer.
+     *
+     * @param usdPrice The new unit price in USD. Fixed point value with 8 decimals.
+     */
+    function setPrice(uint256 usdPrice) external {
+        if (!hasRole(ADMIN_ROLE, msg.sender) && !hasRole(TREASURER_ROLE, msg.sender)) revert Unauthorized();
+        emit SetPrice(usdUnitPrice, usdPrice);
+        usdUnitPrice = usdPrice;
+    }
+
+    /**
+     * @notice Change the maximum supply of storage units. Only callable by admin.
+     *
+     * @param max The new maximum supply of storage units.
+     */
+    function setMaxUnits(uint256 max) external onlyAdmin {
+        /* 1.6e7 = 64TB @ 4MB/unit */
+        if (max > 1.6e7) revert InvalidMaxUnits();
```

Comment:
should we just set this to a much higher number (1e10) incase we ever decide to subdivide the units further?